[PlayListPlayer] Fixed wrong player playlist with playlist files

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
The old PR https://github.com/xbmc/xbmc/pull/11747
has changed the way in how the playlist type is set
by moving all code in one method: `CPlayListPlayer::Play` where now the type is detected by using `IsAudio`/`IsVideo `methods
but this change however, led to a regression with playlist files like STRM, M3U, ...

the strm is an extension set in both audio and video extensions lists,
therefore when you test for `IsAudio`/`IsVideo `this will return true on both methods:
https://github.com/xbmc/xbmc/blob/master/xbmc/PlayListPlayer.cpp#L260-L273
since both methods return true, only the first one is taken in account `IsAudio`
therefore set a wrong playlist, with further repercussions
on the JSON-RPC calls by returning missing/wrong data due to wrong playlist set
such as:
`Player.GetProperties`
`Player.GetActivePlayers`
becoming unusable for addons from first version of Kodi 19 (not sure about Kodi 18)

Then this PR restore the old hint value for playlist type that has been removed,
and that will be used as fallback only when `IsAudio`/`IsVideo` both return true as happens with strm

**A secondary use case fix**
is when you use `Play from here` context menu when you have opened a STRM file that contains more than one links

in this case the video is played from `CGUIMediaWindow::OnPlayAndQueueMedia`
but the playlist set/returned from here https://github.com/xbmc/xbmc/blob/master/xbmc/windows/GUIMediaWindow.cpp#L1512 is wrong. Use `Play from here` from a multiple links STRM is not so popular, therefore is possible that no one complained over the time, this code has never changed in more than 10 years, this is the reason of my changes in `xbmc/view/GUIViewState.cpp`

If possible need to be backported

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Fix #17915

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
call JSON RPC `Player.GetActivePlayers`  and see the player id when you play a strm video must be "1"
call JSON RPC `Player.GetProperties` and check if `currentsubtitle`, `currentvideostream`, `subtitles` data fields exists
more info in the Issue

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Have a working JSON RPC response when you play videos with STRM/M3U files

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

enen92 reviewed (commented):
Thinking a bit about this, instead of reintroducing the argument (which is pretty ugly if you ask me) why not use a listitem property on the item itself (if not already defined) and use that for decision later? Or even keeping everything as is he current and use the property to override the behaviour...
I believe this mostly affects plugin content right? Or playlists resolving to plugin:// VFS paths.

CastagnaIT commented:
take in consideration that this problem will happens for each file extension that has been added (or will be added) on both extensions lists (audio / video) https://github.com/xbmc/xbmc/blob/master/xbmc/settings/AdvancedSettings.cpp#L438-L439

so another example is `.pls` extension, is set on both lists, and will have same problem

then if we add a new listitem property, there are some critical points:

- The file type must have the possibility to add the new property e.g. `#KODIPROP:PlaylistType=video` as text inside the file itself, for STRM/M3U case its ok, but other file format types may not be so (.pls ?).
- I have the suspect that the loading of STRM/M3U file content happens after all this stage, need to be checked, if so it cannot be done so easily
- Many users have huge catalogues of STRM files (tvshows + each episode of each seasons ...etc, hundreds of files and more) therefore each addon-dev have to apply a change to all the hundreds of files of each kodi user, which is not acceptable, each addon will have to create a script to modify all the STRM/M3U of each user to add this new property, so this could be ok if we want to make a transition for Kodi 20 (but this lead to last point)
- This will be a kind of new feature to workaround a regression, but for kodi 20, a backport will be not allowed and Kodi 19 and still be broken, i would like a common solution that can be improved

IMO i would be inclined to still include this regression fix that reitroduce the old behaviour,
this will give the possibility to fix Kodi 19,
then for Kodi 20 we can think to add a new property that allow override the playlist, if it is possible to do so...

enen92 commented:
Sure but still...the item is propagated throughout the chain, you could well define a property on the playlist fileitem to provide the hint without having to overload existing method implementations.

CastagnaIT commented:
i digged more, so manage a "public" override listitem proprerty (#KODIPROP) is not so easy, need to be reworked a lot of things to make playlist created in a different way

however i have removed the overload as suggested, and also i found another bug for a similar use case with context menu, i updated the PR description so you can be read above

enen92 commented on xbmc/video/windows/GUIWindowVideoBase.cpp:
```diff
@@ -1184,11 +1184,13 @@ bool CGUIWindowVideoBase::OnPlayAndQueueMedia(const CFileItemPtr& item, const st
   return CGUIMediaWindow::OnPlayAndQueueMedia(movieItem, player);
 }
 
-void CGUIWindowVideoBase::PlayMovie(const CFileItem *item, const std::string &player)
```
this is a bad sign. Play should play and keep the item as is (so const). Please check where this is called and set the property there

CastagnaIT replied:
done

enen92 commented on xbmc/PlayListPlayer.cpp:
```diff
@@ -258,13 +258,23 @@ bool CPlayListPlayer::PlaySongId(int songId)
+  {
+    // If an extension is set in both audio / video lists (e.g. playlist .strm),
+    // is not possible detect the type of playlist then we rely on the hint
+    playlist = pItem->GetProperty("playlist_type_hint").asInteger32();
```
probably this needs to be validated before calling `.asInteger32()`? I.e. plugin authors may set that property to some random string. What happens in that case?
I've not checked the implementation of the `.asInteger32()` though so it might be fine already. Please double check if possible.

CastagnaIT replied:
i checked and yes its safe, so if the property is set with data other than integer, then the method fallback to default value,
so i have add the default value to PLAYLIST_NONE

enen92 replied:
I'd say that if a plugin resolves its url to a local or remote playlist (e.g. strm) yes, the property will still be set from python although the dynpath changed

enen92 commented on xbmc/view/GUIViewState.cpp:
```diff
@@ -101,7 +101,15 @@ CGUIViewState* CGUIViewState::GetViewState(int windowId, const CFileItemList& it
+    // Playlists (like .strm) can be music or video type
+    // we set the playlist type based on the current
+    // GUI NAV window, otherwise fallback to general
+    if (windowId == WINDOW_VIDEO_NAV)
```
maybe we need a sane default here? What if the window ID is different from both?

CastagnaIT replied:
add the "else" case

enen92 commented on xbmc/view/GUIViewState.cpp:
```diff
@@ -101,7 +101,17 @@ CGUIViewState* CGUIViewState::GetViewState(int windowId, const CFileItemList& it
+    else if (windowId == WINDOW_MUSIC_NAV)
+      return new CGUIViewStateMusicPlaylist(items);
+    else
+      return new CGUIViewStateGeneral(items);
```
why general? shouldn't it be `return new CGUIViewStateMusicPlaylist(items);` just like before?

CastagnaIT replied:
yes changed

CastagnaIT committed:
[PlayListPlayer] Fixed wrong player playlist with playlist files

enen92 reviewed (approved):
Looks good to me and shouldn't cause regressions on jsonrpc like my previous attempt. Nice job

enen92 commented:
jenkins build please

enen92 commented:
jenkins build this please

matthuisman commented:
Just curious, would if make more sense to check for video before audio? I think it's more likely someone playing video vs audio? 

Ie. If we didn't have the hint and it was a playlist, it would check isvideo before isaudio and use video playlist.

lrusak commented:
jenkins build this please

CastagnaIT commented:
the hint prop is always set when needed, have you found a code path where is not set?
